libs/vkd3d/command.c

```diff
@@ -2671,96 +2645,76 @@ static void d3d12_command_list_clear_attachment_pass(struct d3d12_command_list *
-    dependencies[1].dstStageMask = stages;
-    dependencies[1].srcAccessMask = access;
-    dependencies[1].dstAccessMask = 0;
-    dependencies[1].dependencyFlags = 0;
-    dependencies[1].viewOffset = 0;
-
-    pass_info.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO_2_KHR;
-    pass_info.pNext = NULL;
-    pass_info.flags = 0;
-    pass_info.attachmentCount = 1;
-    pass_info.pAttachments = &attachment_desc;
-    pass_info.subpassCount = 1;
-    pass_info.pSubpasses = &subpass_desc;
-    pass_info.dependencyCount = ARRAY_SIZE(dependencies);
-    pass_info.pDependencies = dependencies;
-    pass_info.correlatedViewMaskCount = 0;
-    pass_info.pCorrelatedViewMasks = NULL;
-
-    if ((vr = VK_CALL(vkCreateRenderPass2KHR(list->device->vk_device, &pass_info, NULL, &vk_render_pass))) < 0)
-    {
-        WARN("Failed to create Vulkan render pass, vr %d.\n", vr);
-        return;
     }
 
-    if (!d3d12_command_allocator_add_render_pass(list->allocator, vk_render_pass))
+    memset(&image_barriers[0], 0, sizeof(image_barriers[0]));
+    image_barriers[0].sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    image_barriers[0].srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    image_barriers[0].dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    image_barriers[0].image = resource->res.vk_image;
+    image_barriers[0].newLayout = attachment_info.imageLayout;
+    image_barriers[0].subresourceRange.aspectMask = view->format->vk_aspect_mask;
+    image_barriers[0].subresourceRange.baseMipLevel = view->info.texture.miplevel_idx;
+    image_barriers[0].subresourceRange.levelCount = 1;
+    image_barriers[0].subresourceRange.baseArrayLayer = view->info.texture.layer_idx;
+    image_barriers[0].subresourceRange.layerCount = view->info.texture.layer_count;
+
+    if (resource->desc.Dimension == D3D12_RESOURCE_DIMENSION_TEXTURE3D)
     {
-        WARN("Failed to add render pass.\n");
-        VK_CALL(vkDestroyRenderPass(list->device->vk_device, vk_render_pass, NULL));
-        return;
+        image_barriers[0].subresourceRange.baseArrayLayer = 0;
+        image_barriers[0].subresourceRange.layerCount = 1;
     }
 
-    extent.width = d3d12_resource_desc_get_width(&resource->desc, view->info.texture.miplevel_idx);
-    extent.height = d3d12_resource_desc_get_height(&resource->desc, view->info.texture.miplevel_idx);
-    extent.depth = view->info.texture.layer_count;
+    if (separate_ds_layouts)
+    {
+        image_barriers[0].subresourceRange.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
+
+        image_barriers[1] = image_barriers[0];
+        image_barriers[1].newLayout = stencil_attachment_info.imageLayout;
+        image_barriers[1].subresourceRange.aspectMask = VK_IMAGE_ASPECT_STENCIL_BIT;
+        image_barrier_count = 2;
+    }
+    else
+        image_barrier_count = 1;
 
-    if (!d3d12_command_list_create_framebuffer(list, vk_render_pass, 1, &view->vk_image_view, extent, &vk_framebuffer))
+    for (i = 0; i < image_barrier_count; i++)
```

Comment:
This seems like it would pessimize attachments which don't need layout transition. I would imagine that if oldLayout == newLayout, we can elide barrier for that case. That is what we did implicitly with the older method with external subpass barrier.

Reply:
fixed, sort of. In practice we'll still be doing `UNDEFINED`->`OPTIMAL` for now.

Reply:
Sure. We're kinda force to do UNDEFINED -> OPTIMAL in many clear scenarios due to dumb alias rules + initial resource layout shenanigans. I guess there's a theoretical improvement for later if we deduce:
- Initial layout transition has already happened
- Committed resource